### DISCUSSION TITLE
Should quadratic calculations run onchain or offchain?

### DISCUSSION THREAD

**Discussion by nfrgosselin** [gitcoinco/grants-stack#337] [Q&A] [answered]
As we begin to create the first two fund allocation mechanisms for the protocol (v1 Quadratic Funding and v1 Quadratic Voting), we need to decide whether the calculations should run onchain (where they will be auditable but expensive) or offchain (where they could still be auditable but less decentralized / less expensive). Our default assumption is that they should run offchain, but we're curious if there are strong arguments / alternatives for running them onchain?

**Comment by thelostone-mc** [comment]
Different voting strategy have different means of determining how the pot should be split.
IMO the contract should be responsible for casting a vote and not actually doing the tally itself.

Using contract as a means of storing derived values isn't what we should use it for and something like QF which is computationally expensive. I would recommend going with a centralized node server that can be run independently.

How it might work :
This service would offer an API to
- fetch the latest calculation 
- fetch the latest prediction 

The service would smart enough to look at the round factory and automatically trigger calculations on a periodic basis

And for verification , one can always submit the hash and have other people run the calculations locally / other means to validate the output

**Comment by thelostone-mc** [comment, marked as answer]
The decision here is to run this off-chain via centralised server but this would be an opensource repo which folks can choose to run on their own if they don't trust the system/ if they want to verify it themselves
